kv/txn_coord_sender.go

```diff
@@ -507,72 +598,127 @@ func (tc *TxnCoordSender) updateResponseTxn(argsHeader *proto.RequestHeader, rep
 
 // cleanupTxn is called to resolve write intents which were set down over
 // the course of the transaction. The txnMetadata object is removed from
-// the txns map.
-func (tc *TxnCoordSender) cleanupTxn(txn *proto.Transaction, resolved []proto.Key) {
+// the txns map and taken into account for statistics.
+func (tc *TxnCoordSender) cleanupTxn(txn proto.Transaction, resolved []proto.Key) {
 	tc.Lock()
 	defer tc.Unlock()
 	txnMeta, ok := tc.txns[string(txn.ID)]
 	if !ok {
 		return
 	}
-	txnMeta.close(txn, resolved, tc.wrapped, tc.stopper)
-	delete(tc.txns, string(txn.ID))
+
+	// The supplied txn may be newed than the one in txnMeta, which is relevant
+	// for stats.
+	txnMeta.txn = txn
+	tc.unregisterTxnLocked(txnMeta)
+	txnMeta.close(&txn, resolved, tc.wrapped, tc.stopper)
+}
+
+// unregisterTxnLocked idempotently deletes a txnMetadata object from the sender
+// and collects its stats.
+func (tc *TxnCoordSender) unregisterTxnLocked(txnMeta *txnMetadata) {
+	id := string(txnMeta.txn.ID)
+	if _, ok := tc.txns[id]; !ok {
+		return
+	}
+	tc.txnStats.durations = append(tc.txnStats.durations, float64(tc.clock.PhysicalNow()-txnMeta.firstUpdateNanos))
```

Comment:
I'm not seeing a limit anywhere on the size of `durations` and `restarts`. Am I missing it?